sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -539,29 +571,25 @@ object ScalaReflection extends ScalaReflection {
           Invoke(inputObject, "booleanValue", BooleanType)
 
         case other =>
-          throw new UnsupportedOperationException(
-            s"No Encoder found for $tpe\n" + walkedTypePath.mkString("\n"))
+          val fields = getJavaBeanFields(other)
+          if (fields.length > 0) {
+            CreateNamedStruct(fields.flatMap { f =>
+              val fieldName = f.getName
+              val fieldType = classToType(f.getType)
+              val fieldValue = Invoke(inputObject, getterAndSetter(f)._1, dataTypeFor(fieldType))
+              val clsName = getClassNameFromType(fieldType)
+              val newPath = s"""- java bean field (class: "$clsName", name: "$fieldName")""" +:
+                walkedTypePath
+
+              expressions.Literal(fieldName) :: extractorFor(fieldValue, fieldType, newPath) :: Nil
+            })
+          } else {
+            throw new UnsupportedOperationException(
+              s"No Encoder found for $tpe\n" + walkedTypePath.mkString("\n"))
+          }
       }
     }
   }
-}
-
-/**
- * Support for generating catalyst schemas for scala objects.  Note that unlike its companion
- * object, this trait able to work in both the runtime and the compile time (macro) universe.
- */
-trait ScalaReflection {
```

Comment:
I moved the `schemaFor` stuff to `object ScalaReflection`, because of the need of `mirror.runtimeClass` in `getJavaBeanFields`